File: backend/app/controllers/api/v1/auth_controller.rb
class Api::V1::AuthController < Api::V1::BaseController
  skip_before_action :authenticate_user!, only: [:register, :login]

  def login
    @user = User.find_by(email: params[:email])

    if @user&.authenticate(params[:password]) && @user.deleted_at.nil?
      if @user.seller?
        if @user.pending?
          return render json: { error: 'Seller account pending admin approval.' }, status: :forbidden
        end

        if @user.suspended?
          return render json: { error: 'Seller account suspended by admin.' }, status: :forbidden
        end
      end

      token = JsonWebToken.encode(user_id: @user.id)

      render json: {
        login: 'success',
        token: token,
        user: {
          id: @user.id,
          email: @user.email,
          role: @user.role,
        },
      }
    else
      render json: { error: 'Invalid Credentials!' }, status: :unauthorized
    end
  end

  def register
    @user = User.new(user_params)
    @user.role = permitted_public_role(@user.role)
    @user.active = !@user.seller?
    @user.seller_status = @user.seller? ? :pending : :approved

    if @user.save
      token = JsonWebToken.encode(user_id: @user.id)

      render json: {
        message: 'Registration Success!',
        token: token,
      }, status: :created
    else
      render json: {
        errors: @user.errors.full_messages,
      }, status: :unprocessable_entity
    end
  end

  private

  def user_params
    params.require(:user).permit(:email, :password, :password_confirmation, :role)
  end

  def permitted_public_role(role)
    %w[customer seller].include?(role) ? role : 'customer'
  end
end

File: frontend/app/controllers/auth.controller.js
app.controller(
  'AuthController',
  function ($scope, $http, AuthService, $location) {
    $scope.user = {};
    const API_URL = 'http://localhost:3000/api/v1';

    $scope.login = function () {
      AuthService.login($scope.user)
        .then(function (response) {
          AuthService.saveAuthData(response.data.token, response.data.user);
          $location.path(AuthService.getDashboardPath());
        })
        .catch(function () {
          $scope.error = 'Invalid email or password';
        });
    };

    $scope.register = function () {
      $scope.error = null;
      $scope.success = null;

      if ($scope.user.password !== $scope.user.password_confirmation) {
        $scope.error = 'Passwords do not match';
        return;
      }

      $http
        .post(API_URL + '/auth/register', {
          user: $scope.user,
        })
        .then(function () {
          $scope.success = $scope.user.role === 'seller'
            ? 'Seller account created. Wait for admin approval before logging in.'
            : 'Registration successful. Please login.';
          $location.path($scope.user.role === 'seller' ? '/seller/login' : '/login');
        })
        .catch(function (error) {
          $scope.error = error.data?.errors || 'Registration failed';
        });
    };
  },
);
